### PR TITLE
NGPHistoryParser and medtronic600Simulator code complete

### PR DESCRIPTION
**Pull request by pazaan**
NGPHistoryParser and medtronic600Simulator code complete.
Added unit tests and some extra documentation.
Sorry for the uber-commit @gniezen!

### CODE REVIEW

File: lib/medtronic600/medtronic600Simulator.js [hunk 1]
```diff
@@ -73,291 +111,123 @@ class Medtronic600Simulator {
...
+              .with_scheduleName(this.currBasal.suppressed.scheduleName)
+              .with_duration(Date.parse(event.time) - Date.parse(this.currBasal.basalEndTime))
+              .with_conversionOffset(this.currBasal.conversionOffset)
+              .with_timezoneOffset(this.currBasal.timezoneOffset);
+
+            this.currBasal = restoredBasal;
+          } else {
+            // Otherwise, just fix the start time of the new basal event.
+            event
+              .with_time(this.currBasal.basalEndTime)
```
Comment (gniezen) [inline lib/medtronic600/medtronic600Simulator.js]: Setting `time` without using `fillInUTCInfo` usually leads to unexpected side effects.

File: lib/medtronic600/medtronic600Simulator.js [hunk 2]
```diff
@@ -410,7 +280,8 @@ class Medtronic600Simulator {
       this.suspendingEvent = event;
     } else {
       if (type === 'other' && event.payload.alarm_id === 3 && this.currBasal) {
-        debug('Battery out too long at', event.time, ', annotating previous basal at', this.currBasal.time);
+        debug('Battery out too long at', event.time, ', annotating previous basal at',
```
Comment (gniezen) [inline lib/medtronic600/medtronic600Simulator.js]: Oh, this is good! I should add it to the 5/7-series driver too!

File: lib/medtronic600/medtronic600Simulator.js [hunk 3]
```diff
@@ -493,6 +359,12 @@ class Medtronic600Simulator {
             return false;
           }
         }
+      } else if (event.type === 'basal') {
+        // Filter out any basals with no duration. This can happen inside
+        // applySuspendedBasals(), when a basal is suppressed completely by a pump suspend
```
Comment (gniezen) [inline lib/medtronic600/medtronic600Simulator.js]: Don't you want to know when a basal is being suppressed? This sounds like useful info that we may not have had for other pumps, i.e., maybe you don't even have to look up basal schedules if basals are still coming through during a suspend. Or am I misunderstanding what's happening here?

Comment (gniezen) [inline lib/medtronic600/medtronic600Simulator.js]: Ah, I see further down that you may be using these events already? And then just filtering them out here at the end?

File: lib/medtronic600/medtronic600Simulator.js [hunk 4]
```diff
@@ -510,13 +382,122 @@ class Medtronic600Simulator {
     });
   }
 
+  applySuspendedBasals() {
```
Comment (gniezen) [inline lib/medtronic600/medtronic600Simulator.js]: This function is really hard to thoroughly review by just looking at the code. Do we have sufficient test strategies to make sure all of this works as expected?

File: lib/medtronic600/medtronic600Simulator.js (hunk 2 above)
Comment (gniezen) [inline lib/medtronic600/medtronic600Simulator.js]: Haha, oops, just noticed I put it in the 5/7-series simulator.

File: lib/drivers/docs/medtronic600.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    - `[x]` no insulin
+        - `[ ]` needed to infer a suspend (stoppage of all insulin delivery)
+    - `[ ]` low power
+    - `[ ]` no power
+        - `[ ]` needed to infer a suspend (stoppage of all insulin delivery)
+    - `[ ]` occlusion
+        - `[ ]` needed to infer a suspend (stoppage of all insulin delivery)
+    - `[ ]` no delivery
+        - `[ ]` needed to infer a suspend (stoppage of all insulin delivery)
+    - `[ ]` auto-off
```
Comment (pazaan) [inline lib/drivers/docs/medtronic600.md]: I think that there are, but I need to reverse engineer them. I'm not exactly sure how to *make* the pump have these errors...

File: lib/drivers/medtronic600Driver.js [hunk 6]
```diff
@@ -1119,6 +1164,10 @@ class ReadHistoryCommand extends TransmitPacketRequest {
     this.lastPacketSize = payload.readUInt16BE(0x09);
     this.packetsToFetch = payload.readUInt16BE(0x0B);
     this.segments = [];
+    // Prepopulate the segments array with empty objects so we can check for missing segments later.
```
Comment (pazaan) [inline lib/drivers/medtronic600Driver.js]: It could, should, and has ;)

File: lib/medtronic600/NGPHistoryParser.js [hunk 7]
```diff
@@ -618,28 +832,217 @@ class NGPHistoryParser {
...
+    throw new Error('Matching Temp Basal Start for Temp Basal Complete event not found');
+  }
+
   addBtUtcFields(model, timestamp) {
     model.with_deviceTime(sundial.formatDeviceTime(timestamp.toDate()))
       .set('index', timestamp.rtc)
       .set('jsDate', timestamp.toDate());
 
-    this.cfg.tzoUtil.fillInUTCInfo(model, timestamp.toDate());
+    // Can't fill in UTCInfo for a time change records
```
Comment (pazaan) [inline lib/medtronic600/NGPHistoryParser.js]: Good call. 👍

File: lib/medtronic600/NGPHistoryParser.js [hunk 8]
```diff
@@ -618,28 +832,217 @@ class NGPHistoryParser {
...
+      events.push(settingsChange.done());
+    }
+
+    // Now that we're done using currentSettings for cloning, complete it and push to events.
+    events.push(currentSettings.done());
+
+    return this;
+  }
+
+  buildAlarmRecords(events) {
```
Comment (pazaan) [inline lib/medtronic600/NGPHistoryParser.js]: The pump definitely generates a suspend basal event (I've checked that it does this on a set change, and on a battery swap). But I do need to reverse engineer the alarm message so that we can send the events to the backend.

File: lib/medtronic600/medtronic600Simulator.js [hunk 9]
```diff
@@ -73,291 +111,123 @@ class Medtronic600Simulator {
...
 
-              this.currBasal = resumeBasal;
-              return;
-            }
+        if (event.time > this.currBasal.basalEndTime) {
+          // After a temp basal, the next basal segment start event doesn't happen for up to
+          // a minute after the temp basal complete event, leaving small basal data gaps every
+          // time a temp basal finishes.
+          // It's also feasible that we won't get a basal segment start event at all if the user
+          // has cancelled a temp basal, and followed it up immediately with another temp basal.
```
Comment (pazaan) [inline lib/medtronic600/medtronic600Simulator.js]: No, a temp basal must always be cancelled before a new one can be entered.

File: lib/medtronic600/medtronic600Simulator.js (hunk 3 above)
Comment (pazaan) [inline lib/medtronic600/medtronic600Simulator.js]: Yes, that's right. I did originally start putting the code in `applySuspendedBasals()`, however when I put `_.remove(this.events, event => event === activeSuspendedBasal);` in `applySuspendedBasals()`, eslint complained about [functions in loops](being bad). Perhaps we feel that it's for the best - if you agree, I can move it there.

File: lib/medtronic600/medtronic600Simulator.js (hunk 4 above)
Comment (pazaan) [inline lib/medtronic600/medtronic600Simulator.js]: Yes, I agree - and this is *after* my refactoring! The code that handled suspends was originally in `basal()`, and believe it or not, was *much* more complex than what is now in `basal()` and `applySuspendedBasals()`.
The refactoring came about as a result of me adding a *lot* of unit tests for various simulator scenarios (all based on real event data from Caty's pump). I feel confident that I have all scenarios covered by unit tests (the failures below are because of `node` version incompatibilities, not actual unit tests failures - which I still have to fix).
If you feel that the code is still not maintainable, I can have another look into refactoring it further.

File: lib/medtronic600/NGPHistoryParser.js [hunk 10]
```diff
@@ -545,25 +761,31 @@ class NGPHistoryParser {
...
+      if (_.includes(eventTypes, event.eventType)) {
+        yield event;
       }
     }
+  }
+
+  // This is here because a Temp Basal with a percentage requires the suppressed basal
+  // to determine the rate. Though this could also be done in the simulator, we do it here
+  // because the event is incomplete without the rate.
+  findSuppressedBasal(tempBasal) {
```
Comment (pazaan) [inline lib/medtronic600/NGPHistoryParser.js]: Yes, I've confirmed that basal start events always occur after a temp basal has completed, or in the middle of a temp basal if the schedule changes.

File: lib/drivers/docs/medtronic600.md (hunk 5 above)
Comment (gniezen) [inline lib/drivers/docs/medtronic600.md]: I think Auto-off is now called "auto suspend" on 600-series. Insulin Settings -> Auto Suspend. for "no delivery", block the tubing and/or empty the reservoir. "Over limit" I haven't been able to trigger manually, for "no power" you should ask Eric where to get really poor quality batteries ;)

File: lib/medtronic600/NGPHistoryParser.js (hunk 8 above)
Comment (gniezen) [inline lib/medtronic600/NGPHistoryParser.js]: What I meant is that it may not generate a suspend basal event after a `no_delivery` or `auto_off`.

File: lib/medtronic600/medtronic600Simulator.js (hunk 4 above)
Comment (gniezen) [inline lib/medtronic600/medtronic600Simulator.js]: I didn't mean to imply your code here is complicated or unmaintainable :) I realize that this part of the simulator is inherently complex, so I don't think further refactoring will help. I just wanted to note that I didn't inspect the code in a whole lot of detail as I don't think it will be all that useful, and that a combination of unit tests and manual test strategies will do a better job of making sure it works. 

Do you think it will be possible for Eric to generate some of the edge cases and test scenarios that you discovered on Caty's pump, so that we can check the box that we manually verified them too?

File: lib/medtronic600/NGPHistoryParser.js (hunk 10 above)
Comment (gniezen) [inline lib/medtronic600/NGPHistoryParser.js]: Oh, that's really nice! They seem to be learning from their mistakes :)

File: lib/medtronic600/medtronic600Simulator.js (hunk 4 above)
Comment (pazaan) [inline lib/medtronic600/medtronic600Simulator.js]: I didn't take offence :) I think we're saying the same thing. It's definitely a good idea to give Eric the use cases so that he can generate the data for his own tests.

File: lib/medtronic600/NGPHistoryParser.js [hunk 11]
```diff
@@ -618,28 +832,217 @@ class NGPHistoryParser {
...
+                isf.amount = Math.ceil(isf.amount * 18);
+              }
+            }
+            settingsChange.with_insulinSensitivity(isfs);
+          }
+          break;
+        case NGPHistoryEvent.EVENT_TYPE.NEW_BOLUS_WIZARD_INSULIN_TO_CARB_RATIOS:
+          {
+            const carbRatios = event.carbRatios();
+            // TODO - Exchange to grams conversion
```
Comment (pazaan) [inline lib/medtronic600/NGPHistoryParser.js]: In this particular instance, the data model doesn't support exchanges (see http://developer.tidepool.io/data-model/device-data/types/pumpSettings.html#units), but I've fixed up the mg/dL and mmol/L settings changes.

File: lib/medtronic600/medtronic600Simulator.js (hunk 4 above)
Comment (gniezen) [inline lib/medtronic600/medtronic600Simulator.js]: 👍

File: lib/medtronic600/NGPHistoryParser.js (hunk 11 above)
Comment (gniezen) [inline lib/medtronic600/NGPHistoryParser.js]: Ah, yes. We still have to decide whether we're going to amend the data model, or just use the pump settings that were active at that time to determine whether there are carbs or exchanges in the wizard record: See @jebeck 's latest comment on https://trello.com/c/BnzotMo8.

**Comment by pazaan** [comment]
[Medtronic 6 History Message - Basals <ux:0 dev:3 tst:5>](https://trello.com/c/YM1TDo7Q/105-medtronic-6-history-message-basals-ux-0-dev-3-tst-5)

File: lib/drivers/docs/medtronic600.md [hunk 12]
```diff
@@ -17,7 +17,7 @@
   - `[x]` scheduled basal
     - `[x]` basal rate intervals with a start time, duration, and rate delivered
     - `[x]` name of basal schedule on each scheduled basal rate interval
-    - `[?]` if basal schedule is a single (flat) rate all day, pump records a new basal rate interval every midnight
+    - `[x]` if basal schedule is a single (flat) rate all day, pump records a new basal rate interval every midnight
```
Comment (gniezen) [inline lib/drivers/docs/medtronic600.md]: Oh, does it split up a flat basal rate automatically? That's pretty cool, not many pumps do that.

File: lib/drivers/docs/medtronic600.md [hunk 13]
```diff
@@ -45,18 +45,18 @@ Device-specific? (Add any device-specific notes/additions here.)
     - `[x]` duration of insulin delivery
     - `[x]` amount of insulin delivery programmed (if differs from actual delivery, in case of bolus interruption, cancellation, etc.)
     - `[x]` duration of insulin delivery programmed (if differs from actual duration, in case of bolus interruption, cancellation, etc.)
-    - `*[ ]` extended bolus that crosses midnight is split into two records
+    - `[x]` extended bolus that crosses midnight is split into two records
```
Comment (gniezen) [inline lib/drivers/docs/medtronic600.md]: So the pump does this automatically?

File: lib/medtronic600/NGPHistoryParser.js (hunk 8 above)
Comment (gniezen) [inline lib/medtronic600/NGPHistoryParser.js]: Have you checked this or is it still on your todo list?

File: lib/medtronic600/NGPHistoryParser.js [hunk 14]
```diff
@@ -1266,11 +1265,12 @@ class NGPHistoryParser {
             .with_expectedDuration(matchingBolus.programmedDuration * sundial.MIN_TO_MSEC);
         }
 
-        this.addBtUtcFields(bolus, event.timestamp);
+        this.addTimeFields(bolus, event.timestamp);
 
         if (event.bolusSource === NGPUtil.NGPConstants.BOLUS_SOURCE.BOLUS_WIZARD) {
+          // Use parseFloat(num.toFixed(3)), because in JavaScript, 5.20 + 5.15 = 10.350000000000001
```
Comment (gniezen) [inline lib/medtronic600/NGPHistoryParser.js]: 😂

File: lib/medtronic600/NGPHistoryParser.js [hunk 15]
```diff
@@ -927,10 +925,12 @@ class NGPHistoryParser {
           {
             const targets = event.bgTargets();
             // Because we read the initial pump settings in mg/dL, always convert settings changes.
+            // When we confirm that we can determine pump BG units correctly in
+            // DeviceCharacteristicsResponse:units(), we can remove the conversion here.
             if (event.units === NGPUtil.NGPConstants.BG_UNITS.MMOL_L) {
               for (const target of targets) {
-                target.high = Math.ceil(target.high * 18);
-                target.low = Math.ceil(target.low * 18);
+                target.high = parseFloat((target.high * 18.01559).toFixed(2));
```
Comment (gniezen) [inline lib/medtronic600/NGPHistoryParser.js]: I forgot to mention that there's now a common function to use: `convertBackToMmol` in `commonFunctions.js`.

Comment (gniezen) [inline lib/medtronic600/NGPHistoryParser.js]: Oh wait, you're converting TO mg/dL. Can you factor that out as a common function too? Or maybe just refactor to use a common constant everywhere?

**Comment by gniezen** [comment]
I've made some more comments, but I think we should get this PR merged asap as it's getting pretty unwieldy, and any changes to be put in a new PR.

File: lib/medtronic600/NGPHistoryParser.js (hunk 8 above)
Comment (pazaan) [inline lib/medtronic600/NGPHistoryParser.js]: Yes, it definitely generates a suspend basal event.
